### PR TITLE
ChunkLayers now stored in std::map [revised]

### PR DESCRIPTION
**Pull request by LogicParrot**
A fixed version of #3153 , which was merged and reverted.

### CODE REVIEW

**Comment by SphinxC0re** [comment]
Looks good to me

File: src/ChunkMap.cpp [hunk 1]
```diff
@@ -1677,9 +1660,10 @@ void cChunkMap::RemoveEntity(cEntity * a_Entity)
 bool cChunkMap::ForEachEntity(cEntityCallback & a_Callback)
 {
 	cCSLock Lock(m_CSLayers);
-	for (cChunkLayerList::const_iterator itr = m_Layers.begin(); itr != m_Layers.end(); ++itr)
+	for (auto & itr : m_Layers)
 	{
-		if (!(*itr)->ForEachEntity(a_Callback))
+		cChunkLayer & Layer = itr.second;
+		if (!Layer.ForEachEntity(a_Callback))
```
Comment (LogicParrot) [inline src/ChunkMap.cpp]: For the record, the problem was that I forgot a `!` in #3163 , this prevented mobs from ticking.

Comment (SphinxC0re) [inline src/ChunkMap.cpp]: I know. Isn't that ridiculous :smile:

**Comment by LogicParrot** [comment]
Merging, but I just noticed another problem in master: `RemoveLayer` is never called! This means that `m_Layer` can only grow. I'll treat this in another PR.

File: src/ChunkMap.cpp [hunk 2]
```diff
@@ -78,23 +64,19 @@ void cChunkMap::RemoveLayer(cChunkLayer * a_Layer)
...
 	// Not found, create new:
-	cChunkLayer * Layer = new cChunkLayer(a_LayerX, a_LayerZ, this, *m_Pool);
-	if (Layer == nullptr)
-	{
-		LOGERROR("cChunkMap: Cannot create new layer, server out of memory?");
-		return nullptr;
-	}
-	m_Layers.push_back(Layer);
-	return Layer;
+	m_Layers.emplace(std::piecewise_construct,
```
Comment (LogicParrot) [inline src/ChunkMap.cpp]: Emplace return an iterator, I can get rid of the m_Layers.find  below, and gain some more performance.

File: src/ChunkMap.cpp (hunk 1 above)
Comment (worktycho) [inline src/ChunkMap.cpp]: Its also ridiculous that it wasn't picked up in CI. We need better testing.

**Pull request merged**